packages/core/src/auth/sso/ssoAccessTokenProvider.ts

```diff
@@ -663,9 +674,52 @@ class WebAuthorization extends SsoAccessTokenProvider {
         const cachedRegistration = await this.cache.registration.load(cacheKey)
 
         if (cachedRegistration && (isExpired(cachedRegistration) || cachedRegistration.flow !== 'web auth code')) {
-            await this.invalidate()
+            await this.invalidate('registrationExpired:WebAuth')
         }
 
         return loadOr(this.cache.registration, cacheKey, () => this.registerClient())
     }
 }
+
+/**
+ * Remembers the reason an SSO session was put in to a "needs reauthentication" state.
+ * The current use is for telemetry. When the user reauths, we want {@link AwsLoginWithBrowser}
+ * to know why it needed to be reauthed.
+ *
+ * The flow is to use `setReason()` to remember why the user was put in to a reauth state,
+ * then upon the next reauth use `getReason()`. Finally, use `clearReason()` if the reauth is
+ * successful.
+ */
+export class ReAuthReasonState {
```

Comment:
is a class the right abstraction for this, or is "auth state" something that can/should be generalized in auth.ts?